### ISSUE TITLE
No templates are listed after upgrade to 2.2 until Refresh is clicked

### ISSUE THREAD

**Issue by davidnuzik** [rancher/rancher#19301]
Version: v2.1.7 -> v2.2.0 upgrade (HA or non-HA, happens in both)

**What kind of request is this (question/bug/enhancement/feature request):**
Bug

**Steps to reproduce (least amount of steps as possible):**
- On v2.1.7 add a catalog: https://aliacs-k8s-ap-southeast-1.oss-ap-southeast-1.aliyuncs.com/app/charts-incubator/ (master branch)
- Upgrade to v2.2.0
- Try to launch a catalog app or multi-cluster app. Pick the aliyun catalog. No apps are listed.
- Click the refresh button (in rancher UI) to manually refresh the catalog

**Result:**
You must manually click the refresh button after upgrade for any of the apps to show up in this catalog.

**Other details that may be helpful:**
Originated from issue rancher/rancher#19206

**Comment by mrajashree** [comment]
@davidnuzik Could we include a different catalog here, since this catalog's templates aren't named correctly...it's the same one that has uppercase characters causing the failure right?

**Comment by westlywright** [comment]
I don't think this should be a UI issue. Rancher should refresh the catalogs as part of the upgrade process.

**Comment by davidnuzik** [comment]
Note, this is NOT reproducible in a test done with v2.1.7 -> v2.1.8 upgrade. This is a new issue.

**Comment by davidnuzik** [comment]
This issue happens regardless if HA or non-HA. Also, I tested again with a chart that does not have the aliyun issues and still reproduced.

**Comment by mrajashree** [comment]
In 2.2 we use the new CRD `catalogtemplates` in place of `templates`. These have to be created for the corresponding `templates` that were generated pre-upgrade. The CatalogController refresh method is triggered on server start and starts creating these. But it takes a few seconds to a minute for the catalogtemplates to start getting created.
This info level log "Starting catalog controller" indicates when the catalog controller starts, and it starts on server start without having to click Refresh. Immediately after that the catalogtemplates/versions are created.
You must have clicked on Launch before any catalogtemplate was created immediately after server start, so no templates show up...and UI is not subscribed to template events so none show up even if `catalogtemplates` start getting created, unless there is an explicit GET request to v3/templates.
This GET request is made only when you come back to this page, or on Refresh.
If on upgrade, you exec into your rancher server container and run `kubectl get catalogtemplates -n cattle-global-data` and see some instances being returned, then if you go to Launch apps you will see the templates without any Refresh.
This doesn't happen on upgrade to 2.1.8 since there we're using the same `templates` resource, so creation of new instances isn't required.
I don't think anything can be done to change this

**Comment by alena1108** [comment]
As per @loganhz, Rancher templates UI is not events driven. It means when the new template gets created, you have to refresh the page (or hit Refresh) in order to fetch them.

**Comment by vincent99** [comment]
Are there change events sent from the backend?

**Comment by mrajashree** [comment]
The templates/versions generation events are being sent on v3/subscribe, checked it by adding a catalog and saw the following messages in dev tools
<img width="1352" alt="Screen Shot 2019-04-12 at 11 26 32 AM" src="https://user-images.githubusercontent.com/10672776/56058358-0705e000-5d16-11e9-8d1f-d483982bfed5.png">

But the templates don't show up immediately. They show up on page reload, so as per discussion with @westlywright it's a frontend issue, assigning it over

**Comment by davidnuzik** [comment]
Likely related: #19586

**Comment by davidnuzik** [comment]
Refreshing the UI after an upgrade would require API changes so UI can see that we upgraded rancher. See https://github.com/rancher/rancher/issues/19586#issuecomment-482684848

**Comment by mrajashree** [comment]
But I don't get how this is related to catalogtemplate generation. That happens from catalog controller post refresh without refreshing UI

**Comment by davidnuzik** [comment]
You're probably right @mrajashree I was making a guess that it is related to that issue. I may be completely wrong

**Comment by westlywright** [comment]
QA: I found this quite troublesome to test as it seems to be closely tied to timing and in order to see this happen I had to execute the steps one right after the other and view the catalog index page as fast as I could before the catalog controller has a chance to run through all of its upgrade init jobs. My guess is if you use a catalog with a MASSIVE amount of templates or you had a large amount of resources that require action on upgrade you'd be able to repro this quite easily. Regardless, as new catalog items are added to the catalog after upgrade they will live add on the catalog index page.

```
### Catalog Info
Repo: https://github.com/leodotcloud/rancher-charts
Branch: authelia

### Docker Steps
docker stop <container-id>
docker create --volumes-from <container-id> --name rancher-data rancher/rancher:v2.1.9
docker run --volumes-from rancher-data -v $PWD:/backup alpine tar zcvf /backup/rancher-data-backup-2.1.9-5-6-2019.tar.gz /var/lib/rancher
docker run -d --volumes-from rancher-data --restart=unless-stopped -p 80:80 -p 443:443 rancher/rancher:master
```

**Comment by davidnuzik** [comment]
I am no longer observing the issue.
I tested a couple of times with a couple of different catalogs and was not able to reproduce the issue any longer.

Sometimes for a brief time no catalogs are shown, but they are loaded in once the backend is done with catalog sync now. I was tailing my logs while testing and once this is done everything shows up on the launch app page:
```
2019/05/20 19:19:41 [INFO] Updating global catalog system-library
2019/05/20 19:19:41 [INFO] Catalog sync done. 0 templates created, 3 templates updated, 0 templates deleted
2019/05/20 19:19:41 [INFO] Catalog sync done. 12 templates created, 0 templates updated, 0 templates deleted
2019/05/20 19:19:43 [INFO] Catalog sync done. 31 templates created, 0 templates updated, 0 templates deleted
2019/05/20 19:19:43 [INFO] Catalog sync done. 31 templates created, 0 templates updated, 0 templates deleted
```

Once this is done (which is fairly quick post-upgrade) all my apps show up now without any refresh required either via the Refresh button in Rancher UI or browser refresh.

**Comment by xinghui-hou** [comment]
> In 2.2 we use the new CRD `catalogtemplates` in place of `templates`. These have to be created for the corresponding `templates` that were generated pre-upgrade. The CatalogController refresh method is triggered on server start and starts creating these. But it takes a few seconds to a minute for the catalogtemplates to start getting created.
> This info level log "Starting catalog controller" indicates when the catalog controller starts, and it starts on server start without having to click Refresh. Immediately after that the catalogtemplates/versions are created.
> You must have clicked on Launch before any catalogtemplate was created immediately after server start, so no templates show up...and UI is not subscribed to template events so none show up even if `catalogtemplates` start getting created, unless there is an explicit GET request to v3/templates.
> This GET request is made only when you come back to this page, or on Refresh.
> If on upgrade, you exec into your rancher server container and run `kubectl get catalogtemplates -n cattle-global-data` and see some instances being returned, then if you go to Launch apps you will see the templates without any Refresh.
> This doesn't happen on upgrade to 2.1.8 since there we're using the same `templates` resource, so creation of new instances isn't required.
> I don't think anything can be done to change this @davidnuzik cc @alena1108
### How do I manually trigger a refresh command? web-ui is invalid. I encountered a problem.
![image](https://user-images.githubusercontent.com/10824788/71430442-ccf33780-2706-11ea-847d-367390409295.png)

### logs
```
2019/12/25 03:01:42 [ERROR] ClusterController local [cluster-monitoring-handler] failed with : failed to deploy monitoring: catalogTemplate.management.cattle.io "cattle-global-data/system-library-rancher-monitoring" not found
```
